pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -207,6 +369,192 @@ public void syncMaxReadPositionForNormalPublish(PositionImpl position) {
 
     @Override
     public PositionImpl getMaxReadPosition() {
-        return this.maxReadPosition;
+        if (checkIfReady()) {
+            return this.maxReadPosition;
+        } else {
+            return PositionImpl.earliest;
+        }
+    }
+
+    @Override
+    public void run(Timeout timeout) {
+        if (checkIfReady()) {
+            takeSnapshotByTimeout();
+            this.timer.newTimeout(this, takeSnapshotIntervalTime, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    // we store the maxReadPosition from snapshot then open the non-durable cursor by this topic's manageLedger.
+    // the non-durable cursor will read to lastConfirmedEntry.
+    static class TopicTransactionBufferRecover implements Runnable {
+
+        private final PersistentTopic topic;
+
+        private final TopicTransactionBufferRecoverCallBack callBack;
+
+        private Position startReadCursorPosition = PositionImpl.earliest;
+
+        private final SpscArrayQueue<Entry> entryQueue;
+
+        private final AtomicLong exceptionNumber = new AtomicLong();
+
+        // TODO: MAX_EXCEPTION_NUMBER can config
+        private static final int MAX_EXCEPTION_NUMBER = 500;
```

Comment:
What the expected behavior if the exceptions < MAX_EXCEPTION_NUMBER? In this case, if the transaction buffer recovered, does the transaction buffer will lose some transaction state?

Reply:
it won't lose transaction state, because when read fail the read position will not change. it always read the correct position by the cursor.